File: src/network/packet/impl/request/ServerHandshakeRequestPacket.php
<?php

namespace pocketcloud\cloud\network\packet\impl\request;

use pocketcloud\cloud\console\log\CloudLogger;
use pocketcloud\cloud\event\impl\server\ServerPostVerificationEvent;
use pocketcloud\cloud\event\impl\server\ServerVerifyEvent;
use pocketcloud\cloud\network\client\ServerClient;
use pocketcloud\cloud\network\client\ServerClientCache;
use pocketcloud\cloud\network\packet\data\VerifyStatus;
use pocketcloud\cloud\network\packet\impl\response\ServerHandshakeResponsePacket;
use pocketcloud\cloud\network\packet\impl\ServerSyncPacket;
use pocketcloud\cloud\network\packet\RequestPacket;
use pocketcloud\cloud\network\packet\util\PacketData;
use pocketcloud\cloud\server\CloudServerManager;
use pocketcloud\cloud\server\util\ServerStartSnapshot;
use pocketcloud\cloud\server\util\ServerStatus;

final class ServerHandshakeRequestPacket extends RequestPacket {

    public function __construct(
        private string $serverName = "",
        private int $processId = -1,
        private int $maxPlayers = -1,
    ) {}

    public function handle(ServerClient $client): void {
        if (($server = CloudServerManager::getInstance()->get($this->serverName)) !== null && ServerClientCache::getInstance()->getServer($client) === null) {
            ($ev = new ServerVerifyEvent($server))->call();
            if ($ev->isCancelled()) {
                CloudLogger::get()->warn("Denied server handshake request from §b{} §8(§b{}§8)", $this->serverName, $client->getAddress());
                return;
            }

            $post = ServerStartSnapshot::capture();
            $pre = $server->getPreStartSnapshot();
            $elapsed = round($post->capturedAt - $pre->capturedAt, 2);

            CloudLogger::get()->debug(
                "§b{}§r started in §a{}s§r | TPS: §e{}§r→§e{}§r | TickUsage: §e{}%§r→§e{}%§r | CPU: §e{}%§r→§e{}%§r | RAM: §e{}MB§r→§e{}MB",
                $server->getName(), $elapsed,
                round($pre->avgTps, 1), round($post->avgTps, 1),
                round($pre->tickUsage, 1), round($post->tickUsage, 1),
                round($pre->cpuUsage, 1), round($post->cpuUsage, 1),
                round($pre->memoryUsage), round($post->memoryUsage)
            );

            $server->setPostStartSnapshot($post);
            ServerClientCache::getInstance()->add($server, $client);
            CloudLogger::get()->success("The server §b{} §rhas §aconnected §rto the cloud. §8(§rTook §b{}§rs§8)", $server->getName(), round(microtime(true) - $server->getStartTime(), 3));
            $server->getServerData()->setMaxPlayers($this->maxPlayers);
            $server->getServerData()->setProcessId($this->processId);
            $server->setVerifyStatus(VerifyStatus::VERIFIED);
            $server->addToProxies();
            $server->sync();
            new ServerPostVerificationEvent($server)->call();
            $this->sendResponse(new ServerHandshakeResponsePacket(VerifyStatus::VERIFIED), $client);
            $server->setServerStatus(ServerStatus::ONLINE);
        } else {
            CloudLogger::get()->warn("Denied server handshake request from §b{} §8(§b{}§8)§r, duplicate server...", $this->serverName, $client->getAddress());
            $this->sendResponse(new ServerHandshakeResponsePacket(VerifyStatus::DENIED), $client);
        }
    }

    public function decodePayload(PacketData $packetData): void {
        $packetData->readAll($this->serverName, $this->processId, $this->maxPlayers);
    }

    public function getServerName(): string {
        return $this->serverName;
    }

    public function getProcessId(): int {
        return $this->processId;
    }

    public function getMaxPlayers(): int {
        return $this->maxPlayers;
    }

    public static function create(string $serverName, int $processId, int $maxPlayers): self {
        return new self($serverName, $processId, $maxPlayers);
    }
}

File: cloud-core/src/main/java/de/pocketcloud/cloud/network/packet/impl/request/ServerHandshakeRequestPacket.java
package de.pocketcloud.cloud.network.packet.impl.request;

import de.pocketcloud.api.server.VerificationStatus;
import de.pocketcloud.cloud.PocketCloud;
import de.pocketcloud.cloud.console.log.CloudLogger;
import de.pocketcloud.cloud.event.impl.server.ServerPostVerificationEvent;
import de.pocketcloud.cloud.event.impl.server.ServerVerifyEvent;
import de.pocketcloud.cloud.network.client.ServerClient;
import de.pocketcloud.cloud.network.client.ServerClientCache;
import de.pocketcloud.cloud.network.packet.RequestPacket;
import de.pocketcloud.cloud.network.packet.impl.response.ServerHandshakeResponsePacket;
import de.pocketcloud.common.util.NumberUtils;
import de.pocketcloud.network.packet.data.PacketData;
import de.pocketcloud.api.server.ServerStatus;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.jetbrains.annotations.NotNull;

import java.time.Duration;
import java.time.Instant;

@NoArgsConstructor
@Getter
public final class ServerHandshakeRequestPacket extends RequestPacket {

    private String serverName;
    private long processId;
    private int maxPlayers;

    public ServerHandshakeRequestPacket(String serverName, long processId, int maxPlayers) {
        this.serverName = serverName != null ? serverName : "";
        this.processId = processId;
        this.maxPlayers = maxPlayers;
    }

    @Override
    public void handle(@NotNull ServerClient client) {
        var server = PocketCloud.instance().servers().get(serverName).orElse(null);
        if (server == null) return;
        if (PocketCloud.instance().clients().getServer(client).isEmpty()) {
            var ev = new ServerVerifyEvent(server);
            ev.call();
            if (ev.isCancelled()) {
                server.logger().warn("Denied server handshake request by event §8(§b{}§8)", client.address());
                return;
            }

            float elapsed = Duration.between(server.startTime(), Instant.now()).toSeconds();

            PocketCloud.instance().clients().add(server, client);
            CloudLogger.get().success("The server §b{} §rhas §aconnected §rto the cloud. §8(§rTook §b{}§rs§8)", server.name(), elapsed);
            server.data().maxPlayers(maxPlayers);
            server.data().processId(processId);
            server.verifiedTime(Instant.now());
            server.verificationStatus(VerificationStatus.VERIFIED);
            server.addToProxies();
            server.sync();
            new ServerPostVerificationEvent(server).call();
            sendResponse(new ServerHandshakeResponsePacket(VerificationStatus.VERIFIED), client);
            server.setStatus(ServerStatus.ONLINE);
        } else {
            server.logger().warn("Denied server handshake request, duplicate server... §8(§b{}§r§8)", client.address());
            CloudLogger.get().warn("Denied server handshake request from §b{} §8(§b{}§8)§r, duplicate server...", serverName, client.address());
            sendResponse(new ServerHandshakeResponsePacket(VerificationStatus.DENIED), client);
        }
    }

    @Override
    public void decodePayload(PacketData packetData) {
        this.serverName = packetData.readString();
        this.processId = packetData.readLong();
        this.maxPlayers = packetData.readInt();
    }

    public static ServerHandshakeRequestPacket create(String serverName, int processId, int maxPlayers) {
        return new ServerHandshakeRequestPacket(serverName, processId, maxPlayers);
    }
}
